### ISSUE TITLE
Architecture questions?

### ISSUE THREAD

**Issue by S0c5** [judmorenomo/screenshotapp#1]
## Description

Hello @judmorenomo, thanks to apply to Valiu, after review your repository I'd like to ask you some questions related to:

1.  Why did you choose a sync architecture given the constrains in concurrency and the obvios delayed response by the screenshot task? 

2. What another architecture can you use if we want that at least the 99.999% of the requests should be completed successfully. 

Thanks!

**Comment by judmorenomo** [comment]
Hi @S0c5 thanks a lot for your comments and the challenges, to answer your questions:

1. From my experience and what I’ve learned, HTTP requests are often related to sync architectures and even more when the response is needed by the client that made the request (not that it always has to be like that) so even if the endpoint has a considerable response time it’s often more natural and intuitive to think about a sync architecture instead of an async one. Also, for the purpose of the challenge in which requests can arrive from multiple ips it was easier to solve the challenge this way.

2. I’ll start my answer to this question from the last thing that I said in the first answer: let’s assume that we have a front end that does all the request to our backend, in this case we could respond to all requests with an `accepted` while in the back we are processing the request, now for the back we could split the 2 functions that we are using: the one to get the screenshot and the one to upload it. To do this we could use an event based architecture, for this example I’ll mention RabbitMQ which I have used before.

The main idea is to publish and hear events that lets us know that for some request an action has been done, the flow would look something like this:

1. When we receive the request we could publish an event that says that a request has been received (we need an identifier for each request).
2. We will then hear this event and try to get the screenshot as it is the next step, when we get the screenshot we’ll publish another event that says that we have the screenshot and now we should upload it.
3. We will hear the event from the previous step and when we are able to upload it we’ll publish a message with the URL of the image that is the final result for the request.
4. The front end can hear this final events and using the request identifier update the answer for that request.

How does this help us increasing the number of successful requests? An event based architecture allows us for example to configure retries for messages that had errors, for example if the service we use to upload the images is down we can retry to re process the messages later, or even if our service is down the message will still be pending to process so when our service is back up the message will be processed when our service is back up, splitting the 2 requests that we need let’s us to handle these cases for both functions. 

Note that this could also work even if we don’t have a single source of requests, but we would need something that can respond to each client that made the request after we have the final event, I haven’t done this before so It was easier to explain it assuming that we have a front.

**Comment by S0c5** [comment]
Thanks to answer my questions @judmorenomo, i'd like to see how you implement it in rabbitMQ, are you able to complete this change?

**Comment by judmorenomo** [comment]
Yes of course, I’ll deploy an extra service that will be the one listening to the final result (not a front but just a service that logs something when it has the final result) do you agree? @S0c5

**Comment by judmorenomo** [comment]
Now that I think of it I don’t think that an extra service is that useful, but if you want I can add it, WDYT? @S0c5

**Comment by S0c5** [comment]
@judmorenomo you Right not extra service is necessary just modify your service to implement RabbitMQ!
